mage_ai/streaming/sources/nats_js.py

```diff
@@ -0,0 +1,143 @@
+    config_class = NATSConfig
+
+    def __init__(self, config, **kwargs):
+        self.loop = asyncio.new_event_loop()
+        self.thread = threading.Thread(target=self.start_loop, daemon=True)
+        self.thread.start()
+        self.checkpoint_path = kwargs.get('checkpoint_path')
+        super().__init__(config)
+
+    def start_loop(self):
+        asyncio.set_event_loop(self.loop)
+        self.loop.run_forever()
+
+    def stop_loop(self):
+        self.loop.call_soon_threadsafe(self.loop.stop)
+        self.thread.join()
+
+    async def ainit_client(self):
+        try:
+            self.nc = await nats.connect(
+                self.config.server_url,
+                error_cb=self.error_cb,
+                reconnected_cb=self.reconnected_cb,
+                disconnected_cb=self.disconnected_cb,
+                closed_cb=self.closed_cb,
+            )
+            self.js = self.nc.jetstream()
+            await self.js.add_stream(name=self.config.consumer_name, subjects=[self.config.subject])
+            self.psub = await self.js.pull_subscribe(self.config.subject, "mage_psub")
+        except NoServersError as e:
+            self._print(f'Caught NoServersError while connecting to NATS server: {e}')
+        except Exception as e:
+            self._print(f'Caught exception while connecting to NATS server: {e}')
+
+    # Define callback methods
+    async def disconnected_cb(self):
+        self._print('Got disconnected!')
+
+    async def reconnected_cb(self):
+        self._print(f'Got reconnected to {self.nc.connected_url.netloc}')
+
+    async def error_cb(self, e):
+        self._print(f'There was an error: {e}')
+
+    async def closed_cb(self):
+        self._print('Connection is closed')
+
+    async def aclose_client(self):
+        await self.nc.close()
+
+    def init_client(self):
+        # asyncio.run_coroutine_threadsafe(self.ainit_client(), self.loop)
+        future = asyncio.run_coroutine_threadsafe(self.ainit_client(), self.loop)
+        future.result()
+
+    def close_client(self):
+        asyncio.run_coroutine_threadsafe(self.aclose_client(), self.loop)
+
+    def batch_read(self, handler: Callable):
+        self.init_client()
+
+        try:
+            while True:
+                messages = self.fetch_messages()
+                if not messages:
+                    break
+                handler(messages)  # Pass the entire batch to the handler
+        finally:
+            self.close_client()
+            self.stop_loop()
+
+    async def afetch_messages(self):
+        try:
+            msgs = await self.psub.fetch(self.config.batch_size, timeout=self.config.timeout)
+            return [json.loads(msg.data.decode()) for msg in msgs]
+        except nats.errors.TimeoutError:
+            return []
+
+    def fetch_messages(self):
+        future = asyncio.run_coroutine_threadsafe(self.afetch_messages(), self.loop)
+        return future.result()
+
+    def read(self, handler: Callable):
```

Comment:
Is `read` and `batch_read` are the two primary entrypoints for reading messages? If so, what is the semantic difference between them? I ask because a fetch to the consumer could return zero messages, but the next call could return one or more, so if there is a need to do continuous reading, then a `break` shouldn't occur when zero messages are returned.

Reply:
fixed, thanks